src/meta/SessionMan.cpp

```diff
@@ -0,0 +1,94 @@
+                                    int64_t updateTime) {
+    auto iter = globalSessions_.find(addr);
+    // addr not exists
+    if (iter == globalSessions_.end()) {
+        sessionRec sessRec;
+        sessRec[sessionId] = std::make_pair(startTime, updateTime);
+        globalSessions_.emplace(addr, sessRec);
+    } else {
+        auto idMap = iter->second;
+        auto idIter = idMap.find(sessionId);
+        // SessionId not exists
+        if (idIter == idMap.end()) {
+            globalSessions_[addr][sessionId] = std::make_pair(startTime, updateTime);
+        } else {
+        // Update update time
+            if (globalSessions_[addr][sessionId].first == startTime) {
+                globalSessions_[addr][sessionId].second = updateTime;
+            } else {
+                // Previously invalid data
+                globalSessions_[addr][sessionId] = std::make_pair(startTime, updateTime);
+            }
+        }
+    }
+}
+
+
+void SessionMan::removeSession(const std::string& addr, int64_t sessionId) {
+    auto iter = globalSessions_.find(addr);
+    if (iter == globalSessions_.end()) {
+        return;
+    }
+
+    auto idIter = iter->second.find(sessionId);
+    if (idIter != iter->second.end()) {
+        globalSessions_[addr].erase(idIter);
+        if (globalSessions_[addr].empty()) {
+            globalSessions_.erase(addr);
+        }
+    }
+    return;
+}
+
+
+void SessionMan::removeInvalidSession(const std::unordered_set<std::string>& addrs,
+                                      int64_t UpdateTime) {
+    // Remove invalid session
+    decltype(globalSessions_) globalSessions;
+
+    for (auto gsIter = globalSessions_.begin(); gsIter != globalSessions_.end(); ++gsIter) {
+        auto iterAddr = addrs.find(gsIter->first);
+
+        // According to the update time interval, for graphd crashes
+        if (iterAddr == addrs.end()) {
+            for (auto siter = gsIter->second.begin();  siter != gsIter->second.end(); ++siter) {
+                if (siter->second.second + 2 * FLAGS_heartbeat_interval_secs < UpdateTime) {
+                    continue;
+                } else {
+                    globalSessions[gsIter->first][siter->first] =
+                        std::make_pair(siter->second.first, siter->second.second);
+                }
+            }
+        } else {
+            // According to whether the update time is equal to updateTime, for session expires
+            for (auto siter = gsIter->second.begin();  siter != gsIter->second.end(); ++siter) {
+                if (siter->second.second == UpdateTime) {
+                    globalSessions[gsIter->first][siter->first] =
+                        std::make_pair(siter->second.first, siter->second.second);
+                } else {
+                    continue;
+                }
+            }
+        }
+    }
+    globalSessions_ = std::move(globalSessions);
```

Comment:
is there any performance data that time cost compare between recreate a hasmap and modify a hash map?

Reply:
Do you have a better way to delete data multiple times in the map?

Reply:
I do not the use case about this function and how many members in the map. if the just one member in addrs, but each time need iterator all the member of the globalSessions_.
and if the member of globalSessions_ is very much, then we can save a list save member that sorted by siter->second.second